WordPress/src/androidTest/java/org/wordpress/android/util/UploadWorkRequestTest.kt

```diff
@@ -0,0 +1,160 @@
+package org.wordpress.android.util
+
+import android.support.test.InstrumentationRegistry
+import android.support.test.runner.AndroidJUnit4
+import android.util.Log
+import androidx.work.Configuration
+import androidx.work.WorkInfo
+import androidx.work.WorkManager
+import androidx.work.impl.utils.SynchronousExecutor
+import androidx.work.testing.WorkManagerTestInitHelper
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.`is`
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.wordpress.android.WordPressTest
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.modules.DaggerMockedUploadAppComponent
+import org.wordpress.android.modules.FakeLocalDraftUploadStarter
+import org.wordpress.android.ui.uploads.LocalDraftUploadStarter
+import javax.inject.Inject
+
+@RunWith(AndroidJUnit4::class)
+class UploadWorkRequestTest {
+    @Inject lateinit var mockedLocalDraftUploadStarter: LocalDraftUploadStarter
+    lateinit var fakeLocalDraftUploadStarter: FakeLocalDraftUploadStarter
```

Comment:
We should consider using [WorkerFactory](https://developer.android.com/reference/androidx/work/WorkerFactory). Injecting fakes with Dagger can get unwieldy. And it is harder to track where the objects are coming from or what you're getting. With `WorkerFactory`, you can define what you'll be getting. We can also avoid having the injection happening inside the `Worker` class. 

I somehow got it working with this diff on top of this branch: https://gist.github.com/shiki/edc904cd12821c21201e2f5af665e43f. Here's a summary of the changes:

1. [L253](https://gist.github.com/shiki/edc904cd12821c21201e2f5af665e43f#file-worker-diff-L253
) In `AutoUploadWorker`, we add constructor arguments for what we need. The self-injection can be removed.
2. [L272](https://gist.github.com/shiki/edc904cd12821c21201e2f5af665e43f#file-worker-diff-L272) We create a `Factory` subclass which passes the dependencies to `AutoUploadWorker` when it needs to be created. 
3. [L62](https://gist.github.com/shiki/edc904cd12821c21201e2f5af665e43f#file-worker-diff-L62) In `UploadWorkRequestTest`, we create mocks of `LocalDraftUploadStarter` and `SiteStore`. In [L84](https://gist.github.com/shiki/edc904cd12821c21201e2f5af665e43f#file-worker-diff-L84), we create a local `AutoUploadWorker.Factory` and pass the mocked dependencies. When the `AutoUploadWorker` is created, this factory will be used. 
4. [L217](https://gist.github.com/shiki/edc904cd12821c21201e2f5af665e43f#file-worker-diff-L217) For the app classes, we set up a default `AutoUploadWorker.Factory` using the pre-existing dependencies injected in `WordPress.java`.  

The tests still passed which means the mocks worked. I tested it with the app running normally and it seems to be able to create the `AutoUploadWorker` just fine. 

Note: The code in the diff can be better. 😅 

What do you think? Is there something that's missing?

Reply:
I was really confused yesterday why had google introduced WorkerFactory just for Worker subclasses. I thought `ListanableWorker` extends `Worker` but it's the other way round (the naming is confusing, right ?:D).... So we don't even need to refactor it to CoroutinesWorker and we can still use the WorkerFactory ❤️ .

Reply:
I actually started to work on a solution with the WorkFactory but I didn't like that too much as it was adding complexity on the app side. I preferred to keep the app simple and add complexity in tests. Since then I added much complexity with the custom AppComponent/Module/fake class and I didn't reassess the WorkFactory solution. I'll have a look at the patch, but if we can drop the fake class, that's a better solution.

Reply:
Thanks @shiki - fixed in 5be63ca  - I also re-tried the option in Mockito2 to [mock final classes/methods](https://github.com/mockito/mockito/wiki/What's-new-in-Mockito-2#mock-the-unmockable-opt-in-mocking-of-final-classesmethods) but it's still failing in instrumented tests (works well in our unit test).